### DISCUSSION TITLE
Text Extraktion from docx and pptx files

### DISCUSSION THREAD

**Discussion by simonschoe** [pymupdf/pymupdf4llm#102] [Q&A]
Hi there, on the website you state that text can be extracted from all sorts of documents (e.g., docx and pptx): https://pymupdf4llm.readthedocs.io/en/latest/. Are there any examples how I would best proceed if I have docx and/or pptx files instead of PDF files?

**Comment by hewliyang** [comment]
convert them to PDFs

**Comment by JorjMcKie** [comment]
You can directly use them by their filenames like "document.docx".
The issue is that page sizes in these cases are fluid "reflowable", and no tables or text columns are recognized.
It is recommended to therefore regard the full document as one large page, which is the default: height is set to None.

**Comment by simonschoe** [comment]
> You can directly use them by their filenames like "document.docx". The issue is that page sizes in these cases are fluid "reflowable", and no tables or text columns are recognized. It is recommended to therefore regard the full document as one large page, which is the default: height is set to None.

When I try reading the `docx` file directly, I currently obtain an `AssertionError`:
```python
import pymupdf4llm
elements = pymupdf4llm.to_markdown("testfile.docx")
```

```python
...

~\AppData\Roaming\Python\Python311\site-packages\pymupdf\__init__.py in ?(page, required)
    333         return page
    334     elif isinstance(page, mupdf.FzPage):
    335         ret = mupdf.pdf_page_from_fz_page(page)
    336         if required:
--> 337             assert ret.m_internal
    338         return ret
    339     elif page is None:
    340         assert 0, f'page is None'

AssertionError: 
```

**Comment by JorjMcKie** [comment]
I just confirmed correct behavior using pymupdf4llm v0.0.10 and pymupdf v1.24.9.
Transferring this thread to Discussions tab.

**Comment by simonschoe** [comment]
I can confirm that it was an error particular to `pymupdf v1.24.7` (https://github.com/pymupdf/PyMuPDF/issues/3654). Is now fixed.

**Comment by Filimoa** [comment]
I think this is a common use case of wanting to convert xlsx to pdf - not sure the best way to do this with pymupdfpro?

```python
import pymupdf.pro
pymupdf.pro.unlock(KEY)

doc = pymupdf.open("./Collision Extract Layout.xls")
doc.save("./Collision Extract Layout.pdf")
```

Returns 
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
/var/folders/yw/680m_xnx3b9427r40r4hdzv40000gn/T/ipykernel_69191/605626560.py in ?()
      1 doc = pymupdf.open("/Users/sergey/Downloads/Collision Extract Layout.xls")
      2 
      3 # save as pdf
----> 4 doc.save("/Users/sergey/Downloads/aug22.pdf", clean=True)

~/Developer/business/filings-ai-website/backend/.venv/lib/python3.12/site-packages/pymupdf/__init__.py in ?(self, filename, garbage, clean, deflate, deflate_images, deflate_fonts, incremental, ascii, expand, linear, no_new_id, appearance, pretty, encryption, permissions, owner_pw, user_pw, preserve_metadata, use_objstms, compression_effort)
   5620                 raise ValueError("incremental needs original file")
   5621         if user_pw and len(user_pw) > 40 or owner_pw and len(owner_pw) > 40:
   5622             raise ValueError("password length must not exceed 40")
   5623 
-> 5624         pdf = _as_pdf_document(self)
   5625         opts = mupdf.PdfWriteOptions()
   5626         opts.do_incremental = incremental
   5627         opts.do_ascii = ascii

~/Developer/business/filings-ai-website/backend/.venv/lib/python3.12/site-packages/pymupdf/__init__.py in ?(document, required)
    469         return document
    470     elif isinstance(document, mupdf.FzDocument):
    471         ret = mupdf.PdfDocument(document)
    472         if required:
--> 473             assert ret.m_internal
    474         return ret
    475     elif document is None:
    476         assert 0, f'document is None'

AssertionError: 
```

Docs are unhelpful.

**Reply by JorjMcKie** [reply]
> Docs are very unhelpful.

Well, admittedly we haven't yet implemented this telepathy feature 😂!
You still must endeavor entering "convert" in the documentation's search field. Scrolling down a bit leads you to this 

![image](https://github.com/user-attachments/assets/272075fc-6d34-49ff-8adf-859b8d6cfa9d)

The `Document.save()` method's documentation on the other hand clearly says "PDF only".
So, what else could we have done?

**Reply by Filimoa** [reply]
Raise an 'NotImplementedError' instead? I'm not sure your screenshot tells the user anything  about this not being implemented.

**Reply by JorjMcKie** [reply]
As a new user, I would click on the link. Then I would see that **_every_** supported document type can be converted to a PDF.
